types/node/util.d.ts

```diff
@@ -1105,6 +1105,210 @@ declare module 'util' {
+     *
+     *     - `type` Type of argument, which must be either `boolean` (for options
+     *        which do not take values) or `string` (for options which do).
+     *     - `multiple` Whether this option can be provided multiple
+     *       times. If `true`, all values will be collected in an array. If
+     *       `false`, values for the option are last-wins. **Default:** `false`.
+     *     - `short` A single character alias for the option.
+     *
+     *   - `strict`: Whether an error should be thrown when unknown arguments
+     *     are encountered, or when arguments are passed that do not match the
+     *     `type` configured in `options`. **Default:** `true`.
+     *   - `allowPositionals`: Whether this command accepts positional arguments.
+     *     **Default:** `false` if `strict` is `true`, otherwise `true`.
+     *   - `tokens`: Whether tokens {boolean} Return the parsed tokens. This is useful
+     *     for extending the built-in behavior, from adding additional checks through
+     *     to reprocessing the tokens in different ways.
+     *     **Default:** `false`.
+     *
+     * @returns The parsed command line arguments:
+     *
+     *   - `values` A mapping of parsed option names with their string
+     *     or boolean values.
+     *   - `positionals` Positional arguments.
+     *   - `tokens` Detailed parse information (only if `tokens` was specified).
+     *
+     */
+    export function parseArgs<T extends Exact<ParseArgsConfig, T>>(config: T): ParsedResults<T>;
+
+    interface ParseArgsOptionConfig {
+        type: 'string' | 'boolean';
+        short?: string;
+        multiple?: boolean;
+    }
+
+    interface ParseArgsOptionsConfig {
+        [longOption: string]: ParseArgsOptionConfig;
+    }
+
+    interface ParseArgsConfig {
+        strict?: boolean;
+        allowPositionals?: boolean;
+        tokens?: boolean;
+        options?: ParseArgsOptionsConfig;
+    }
+
+    /*
+    IfDefaultsTrue and IfDefaultsFalse are helpers to handle default values for missing boolean properties.
+    TypeScript does not have exact types for objects: https://github.com/microsoft/TypeScript/issues/12936
+    This means it is impossible to distinguish between "field X is definitely not present" and "field X may or may not be present".
+    But we expect users to generally provide their config inline or `as const`, which means TS will always know whether a given field is present.
+    So this helper treats "not definitely present" (i.e., not `extends boolean`) as being "definitely not present", i.e. it should have its default value.
+    This is technically incorrect but is a much nicer UX for the common case.
+    The IfDefaultsTrue version is for things which default to true; the IfDefaultsFalse version is for things which default to false.
+    They differ only in the final item of the if-else chain.
+    */
+    type IfDefaultsTrue<T, IfTrue, IfFalse> = T extends true
```

Comment:
I'm a little worried about the complexity of these type defs, this is some really advanced stuff and I had to read it over about 8 times before I understood what is going on.

I wonder if a simpler approach would be beneficial both for maintainers and consumers.

Reply:
If you can suggest a simpler approach which will make

```ts
let { values } = parseArgs({ options: { x: { type: 'string' } });
let x: string | undefined = values.x;
// @ts-expect-error `y` was not one of the configured options
let y = values.y;
```

typecheck (and similarly for other parts), I'd be very happy to adopt it! Otherwise, though, I think this complexity is worth it for consumers; having less precise types here would be really annoying.

Reply:
I agree with Simon here, thx!

Reply:
If you are willing to limit the complexity for `positionals` and `tokens`, a strongly typed `values` can be achieved far more simply:

```ts
// NOTE: Uses ParseArgsConfig, ParseArgsOptionsConfig, and ParseArgsOptionConfig as defined in PR

type ParseArgsValue<T extends ParseArgsOptionConfig> =
    T extends { multiple: true } ? ParseArgsValue<Omit<T, 'multiple'>>[] :
    T extends { type: 'string' } ? string :
    T extends { type: 'boolean' } ? boolean :
    string | boolean;

type ParseArgsValues<T extends ParseArgsConfig> =
    T extends { strict: true } ?
        { -readonly [P in keyof T['options'] & string]?: ParseArgsValue<NonNullable<T['options']>[P]>; } :
        & Partial<Record<string, ParseArgsValue<any>>
        & Record<keyof T['options'] & string, ParseArgsValue<any>>>;
        // NOTE: `ParseArgsValue<any>` is `string | boolean | (string | boolean)[]`

type ParseArgsToken =
    | { kind: 'option', index: number, name: string, rawName: string, value: string, inlineValue: boolean }
    | { kind: 'option', index: number, name: string, rawName: string, value: undefined, inlineValue: undefined }
    | { kind: 'option-terminator', index: number }
    | { kind: 'positional', index: number, value: string };

type ParseArgsResults<T extends ParseArgsConfig> = {
    values: ParseArgsValues<T>;
    positionals: string[];
    tokens: T extends { tokens: true } ? ParseArgsToken[] : undefined;
};

export function parseArgs<T extends ParseArgsConfig>(config: T): ParseArgsResults<T>;
```